### PR TITLE
CORE-5553 Added The UTXO Token Observer API to support token selection.

### PR DESCRIPTION
**Pull request by owenstanford**
Added the UTXO state observer API used to generate tokens for the token selection cache.

### CODE REVIEW

File: ledger/ledger-utxo-observable/src/main/kotlin/net/corda/v5/ledger/utxo/observer/UtxoToken.kt [hunk 1]
```diff
@@ -0,0 +1,18 @@
+package net.corda.v5.ledger.utxo.observer
+
+import net.corda.v5.ledger.utxo.token.selection.TokenClaimCriteria
+import net.corda.v5.ledger.utxo.token.selection.TokenSelection
+import java.math.BigDecimal
+
+/**
+ * The [UtxoToken] defines the structure of a token produced by an instance of [UtxoLedgerTokenStateObserver]
+ *
+ * @property poolKey The key of the token pool this token belongs to.
```
Comment (mnesbit) [inline ledger/ledger-utxo-observable/src/main/kotlin/net/corda/v5/ledger/utxo/observer/UtxoToken.kt]: Need to state explicitly that all tokens in a pool are regarded as 'fungible' and can be spent together, merged, etc

File: ledger/ledger-utxo-observable/src/main/kotlin/net/corda/v5/ledger/utxo/observer/UtxoTokenFilterFields.kt [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+/**
+ * The [UtxoTokenFilterFields] provide optional fields to allow CorDapps to filter for subsets of tokens within a token
+ * pool.
+ *
+ * @property tag Optional user defined string that can be used for regular expression filters.
+ * @ownerHash Optional token owner hash.
+ */
+data class UtxoTokenFilterFields(
+    val tag: String?,
+    val ownerHash: SecureHash?
```
Comment (mnesbit) [inline ledger/ledger-utxo-observable/src/main/kotlin/net/corda/v5/ledger/utxo/observer/UtxoTokenFilterFields.kt]: For reference right now getting hold of a digest as a user of application is quite hard as it isn't in application/crypto like it should be. We will need a an indication of how to do PublicKey to id (ie. sha256(PublicKey.encoded), but I am hoping we get one on the api for other reasons anyway

File: ledger/ledger-utxo-observable/src/main/kotlin/net/corda/v5/ledger/utxo/observer/UtxoTokenPoolKey.kt [hunk 3]
```diff
@@ -0,0 +1,22 @@
+package net.corda.v5.ledger.utxo.observer
+
+import net.corda.v5.crypto.SecureHash
+
+/**
+ * The [UtxoTokenPoolKey] defines a key for a pool of similar tokens.
```
Comment (mnesbit) [inline ledger/ledger-utxo-observable/src/main/kotlin/net/corda/v5/ledger/utxo/observer/UtxoTokenPoolKey.kt]: Should mention that we also segregate by notary automatically

File: ledger/ledger-utxo-observable/src/main/kotlin/net/corda/v5/ledger/utxo/observer/UtxoToken.kt (hunk 1 above)
Comment (owenstanford) [reply ledger/ledger-utxo-observable/src/main/kotlin/net/corda/v5/ledger/utxo/observer/UtxoToken.kt]: As discussed, will follow up with the UX team about where best to add higher level documentation.

File: ledger/ledger-utxo-observable/src/main/kotlin/net/corda/v5/ledger/utxo/observer/UtxoTokenPoolKey.kt (hunk 3 above)
Comment (owenstanford) [reply ledger/ledger-utxo-observable/src/main/kotlin/net/corda/v5/ledger/utxo/observer/UtxoTokenPoolKey.kt]: updated

File: ledger/ledger-utxo/src/main/kotlin/net/corda/v5/ledger/utxo/observer/UtxoLedgerTokenStateObserver.kt [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+ *     override val stateType =  ExampleState::class.java
+ *
+ *     override fun onProduced(stateAndRef: StateAndRef<ExampleState>): UtxoToken {
+ *         val state = stateAndRef.state.contractState
+ *         return UtxoToken(
+ *             UtxoTokenPoolKey(ExampleState::class.java.typeName, state.issuer, state.currency),
+ *             state.amount,
+ *             UtxoTokenFilterFields()
+ *         )
+ *     }
```
Comment (blsemo) [inline ledger/ledger-utxo/src/main/kotlin/net/corda/v5/ledger/utxo/observer/UtxoLedgerTokenStateObserver.kt]: This claims to be a Java example but looks a bit kotlinesque to me.

**Review by blsemo — COMMENTED** [review]
Looks fine to me apart from a wrong example in the KDoc.
Also, I do we need to document what users have to do with the UtxoLedgerTokenStateObservers they have implemented? Or does the runtime discover them automatically (which probably should be stated explicitly in the KDoc).

File: ledger/ledger-utxo/src/main/kotlin/net/corda/v5/ledger/utxo/observer/UtxoLedgerTokenStateObserver.kt (hunk 4 above)
Comment (owenstanford) [reply ledger/ledger-utxo/src/main/kotlin/net/corda/v5/ledger/utxo/observer/UtxoLedgerTokenStateObserver.kt]: fixed.

**Review by mnesbit — APPROVED** [review]
OK

**Pull request merged**